File: sort/Stack.go
package sortPkg

type StackElement[T any] struct {
	Value T
	Prev  *StackElement[T]
}

type Stack[T any] struct {
	Size int
	Top  *StackElement[T]
}

func (s *Stack[T]) Push(item T) {
	node := StackElement[T]{Value: item}
	s.Size++

	if s.Top == nil {
		s.Top = &node
		return
	}

	node.Prev = s.Top
	s.Top = &node
}

func (s *Stack[T]) Pop() *T {
	if s.Top == nil {
		return nil
	}

	s.Size--
	top := s.Top.Value
	s.Top = s.Top.Prev
	return &top
}

func (s *Stack[T]) Peek() *T {
	return &s.Top.Value
}

File: sort/Stack.ts
type Node<T> = {
    value: T
    prev?: Node<T>
}
export default class Stack<T>{
    public size: number
    public top?: Node<T>

    constructor() {
        this.size = 0
        this.top = undefined
    }

    push(item?: T): void {
        const node = { value: item } as Node<T>
        this.size++
        if (!this.top) {
            this.top = node
            return
        }

        node.prev = this.top
        this.top = node

    }

    pop(): T | undefined {
        this.size = Math.max(this.size - 1, 0)
        if (this.size === 0) {
            const top = this.top?.value
            this.top = undefined
            return top
        }

        const top = this.top?.value
        this.top = this.top?.prev
        return top
    }

    peek(): T | undefined {
        return this.top?.value
    }
} 